src/IO/ISchedulerNode.h

```diff
@@ -97,34 +158,128 @@ class EventQueue
+        {
+            processQueue(lock);
+            return true;
+        }
+        if (!postponed.empty())
+        {
+            processPostponed(lock);
+            return true;
+        }
+        return false;
+    }
+
+    /// Process single event if it exists and meets postponing constraint
     /// Returns `true` iff event has been processed
     bool tryProcess()
     {
         std::unique_lock lock{mutex};
-        if (queue.empty())
+        if (!queue.empty())
+        {
+            processQueue(lock);
+            return true;
+        }
+        if (postponed.empty())
             return false;
-        Event event = std::move(queue.front());
-        queue.pop_front();
-        lock.unlock(); // do not hold queue mutext while processing events
-        event();
-        return true;
+        else
+        {
+            if (postponed.front().key <= now())
+            {
+                processPostponed(lock);
+                return true;
+            }
+            return false;
+        }
     }
 
     /// Wait for single event (if not available) and process it
     void process()
     {
         std::unique_lock lock{mutex};
-        pending.wait(lock, [&] { return !queue.empty(); });
+        while (true)
+        {
+            if (!queue.empty())
+                return processQueue(lock);
+            if (postponed.empty())
+                wait(lock);
+            else
+            {
+                if (postponed.front().key <= now())
+                    return processPostponed(lock);
+                waitUntil(lock, postponed.front().key);
+            }
+        }
+    }
+
+    TimePoint now()
+    {
+        if (auto result = manual_time.load(); likely(result == TimePoint()))
+            return std::chrono::system_clock::now();
+        else
+            return result;
+    }
+
+    /// For testing only
+    void setManualTime(TimePoint value)
+    {
+        std::unique_lock lock{mutex};
+        manual_time.store(value);
+        pending.notify_one();
+    }
+
+    /// For testing only
+    void advanceManualTime(Duration elapsed)
+    {
+        std::unique_lock lock{mutex};
+        manual_time.store(manual_time.load() + elapsed);
+        pending.notify_one();
+    }
+
+private:
+    void wait(std::unique_lock<std::mutex> & lock)
+    {
+        pending.wait(lock);
+    }
+
+    void waitUntil(std::unique_lock<std::mutex> & lock, TimePoint t)
+    {
+        if (likely(manual_time.load() == TimePoint()))
+            pending.wait_until(lock, t);
+        else
+            pending.wait(lock);
+    }
+
+    void processQueue(std::unique_lock<std::mutex> & lock)
+    {
         Event event = std::move(queue.front());
         queue.pop_front();
-        lock.unlock(); // do not hold queue mutext while processing events
+        lock.unlock(); // do not hold queue mutex while processing events
+        event();
+    }
+
+    void processPostponed(std::unique_lock<std::mutex> & lock)
```

Comment:
Maybe use move since caller may not expect that we will unlock it. Or perhaps the fact that it's not a constant already shows that.

```suggestion
    void processPostponed(std::unique_lock<std::mutex> && lock)
```

Reply:
Yes, this is indeed not a super clear place. Let me check, maybe a move will work better.

Reply:
Done